Add basic implementation of a QAbstractGrpcChannel based on gRPC C++

rapgenic opened the pull request:
This is a basic pull request referring to issue #145.

It is still not ready to merge, however if you could give me some feedback on the implementation that would be great!

semlanik commented:
@rapgenic, kindly sorry, I will be able to check commit next week only. Thank you for efforts you done!

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,382 @@
+
+}
+
+QGrpcChannel::QGrpcChannel(const QUrl &url, std::unique_ptr<QAbstractGrpcCredentials> credentials) : QAbstractGrpcChannel()
```
For credentials please use native grpc credentials here directly instead. As per design constructor might be called by used with any parameters, so it's better to use grpc::ChannelCredentials here

semlanik commented:
Also one more thing, use --amend option for commits to make implementation more solid in future.

rapgenic commented:
Hi @semlanik thank you for your review! I am sorry for answering that late, but right now I am on vacation, so I haven't been checking my notifications lately... 

As soon as I am back I will address all the issues you pointed out and I'll try to come up with some tests for the feature.

> Also one more thing, use --amend option for commits to make implementation more solid in future.

I will rebase all the commits into a single one so that the history is cleaner.

rapgenic commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,382 @@
+                status = parseByteBuffer(response, &data);
+
+                if (!status.ok())
+                    goto end;
```
Sorry for that :) That's quite common for C style error handling...

rapgenic commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,382 @@
+
+end:
+            this->status = {
+                (QGrpcStatus::StatusCode) status.error_code(),
```
reinterpret_cast does not work, I think it's static_cast

rapgenic commented:
Hi! Just wanted to update you, I'm working on this and addressing the issues you pointed out, but I do not have much time as of now, so I apologize in advance if I'll be a little slower for the next few days.

rapgenic commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,382 @@
+};
+
+//! \private
+class QGrpcChannelWorker : public QObject {
```
Could you tell me briefly how you would use QFuture or std::future for the worker class, since I have no idea?

FYI my implementation comes from Qt documentation, which suggests creating a Worker class and calling moveToThread()

semlanik replied:
Future it self is already implementer worker thread. So idea is to create future and wait until opration is finished, either in same stack or subscribe to it finish. Overall we may keep current implementation for now

semlanik commented:
Hi @rapgenic , take your time. I'm not hurry and will be glad to help if you have any questions.

rapgenic commented:
Hi @semlanik I am working on the tests (with some difficulty :) ), and I am not able to build them with the following error message:

```
[ 36%] Linking CXX executable qtprotobuf_qml_test
[ 99%] Built target qtprotobuf_test_multifile_qtprotobuf_gen
[100%] Built target qtprotobuf_test_multifile
/usr/bin/ld: CMakeFiles/qtgrpc_qml_test.dir/main.cpp.o: in function `TestSetup::TestSetup()::StaticQtProtobufQuickPluginPluginInstance::StaticQtProtobufQuickPluginPluginInstance()':
main.cpp:(.text._ZZN9TestSetupC4EvEN41StaticQtProtobufQuickPluginPluginInstanceC2Ev[_ZZN9TestSetupC4EvEN41StaticQtProtobufQuickPluginPluginInstanceC5Ev]+0xe): undefined reference to `qt_static_plugin_QtProtobufQuickPlugin()'
/usr/bin/ld: CMakeFiles/qtgrpc_qml_test.dir/main.cpp.o: in function `TestSetup::TestSetup()::StaticQtGrpcQuickPluginPluginInstance::StaticQtGrpcQuickPluginPluginInstance()':
main.cpp:(.text._ZZN9TestSetupC4EvEN37StaticQtGrpcQuickPluginPluginInstanceC2Ev[_ZZN9TestSetupC4EvEN37StaticQtGrpcQuickPluginPluginInstanceC5Ev]+0xe): undefined reference to `qt_static_plugin_QtGrpcQuickPlugin()'
/usr/bin/ld: CMakeFiles/qtgrpc_qml_test.dir/main.cpp.o: in function `TestSetup::TestSetup()':
main.cpp:(.text._ZN9TestSetupC2Ev[_ZN9TestSetupC5Ev]+0x96): undefined reference to `qt_static_plugin_QtProtobufQuickPlugin()'
/usr/bin/ld: main.cpp:(.text._ZN9TestSetupC2Ev[_ZN9TestSetupC5Ev]+0x10a): undefined reference to `qt_static_plugin_QtGrpcQuickPlugin()'
collect2: error: ld returned 1 exit status
make[2]: *** [tests/test_grpc_qml/CMakeFiles/qtgrpc_qml_test.dir/build.make:168: tests/test_grpc_qml/qtgrpc_qml_test] Errore 1
make[1]: *** [CMakeFiles/Makefile2:2497: tests/test_grpc_qml/CMakeFiles/qtgrpc_qml_test.dir/all] Errore 2
make[1]: *** Attesa per i processi non terminati....
/usr/bin/ld: CMakeFiles/qtprotobuf_qml_test.dir/main.cpp.o: in function `TestSetup::TestSetup()::StaticQtProtobufQuickPluginPluginInstance::StaticQtProtobufQuickPluginPluginInstance()':
main.cpp:(.text._ZZN9TestSetupC4EvEN41StaticQtProtobufQuickPluginPluginInstanceC2Ev[_ZZN9TestSetupC4EvEN41StaticQtProtobufQuickPluginPluginInstanceC5Ev]+0xe): undefined reference to `qt_static_plugin_QtProtobufQuickPlugin()'
/usr/bin/ld: CMakeFiles/qtprotobuf_qml_test.dir/main.cpp.o: in function `TestSetup::TestSetup()':
main.cpp:(.text._ZN9TestSetupC2Ev[_ZN9TestSetupC5Ev]+0x87): undefined reference to `qt_static_plugin_QtProtobufQuickPlugin()'
collect2: error: ld returned 1 exit status
make[2]: *** [tests/test_qml/CMakeFiles/qtprotobuf_qml_test.dir/build.make:171: tests/test_qml/qtprotobuf_qml_test] Errore 1
make[1]: *** [CMakeFiles/Makefile2:2556: tests/test_qml/CMakeFiles/qtprotobuf_qml_test.dir/all] Errore 2
make: *** [Makefile:182: all] Errore 2
```

Can you give me any suggestion?

rapgenic commented:
Actually it seems to happen on both versions, mine and yours (master) when I run make a second time...

semlanik commented:
Probably it's related to version conflicts between QtProtobuf in your system and one in build directory. Sometimes it happens.

Required symbols should be in libprotobufquickplugin.a and could be easly checked by this command:
```
objdump -x src/protobuf/quick/libprotobufquickplugin.a | grep qt_static_plugin_QtProtobufQuickPlugin
```
Same applicable for grpc.

rapgenic commented:
Hi @semlanik, finally I think I have managed to extend the test suite to check the grpc backend.

I extended the tests corresponding to the two folders `test_grpc` and `test_grpc_qml`:

* For the tests in `test_grpc_qml` I simply copied the main file with a small modification to the class in test.h to allow me to pass the right channel type for testing while keeping the same class.
* For the ssl test in `test_grpc` I copied the single test, as it was very simple.
* For the main test file `clienttest.cpp` I converted most of the tests to parametrized tests, which are run three times, each one with a client to which is attached a different channel for every run. This allows to write all the tests only once but to be valid for every backend.

Note that the tests only run when CMake variable `QT_PROTOBUF_NATIVE_GRPC_CHANNEL=ON`.

When you have some time could you give me your feedback?
Also why the CI is failing on building the docker image??

P.s.: In the meantime as the tests were giving me a few errors, I cleaned and simplified the code, so you might want to have a look at it too.

semlanik commented:
Hi @rapgenic ,
I will check your commits shortly. About CI fail, you may check logs. Actually "build docker image" is only rule name. There is compilation issue
```
/build/tests/test_grpc_qml/../../../sources/tests/test_grpc_qml/test.h:54:10: error: no matching function for call to 'qmlRegisterSingletonType<qtprotobufnamespace::tests::TestServiceClient>(const char [26], int, int, const char [18], TestSetup::TestSetup(std::shared_ptr<QtProtobuf::QAbstractGrpcChannel>)::<lambda(QQmlEngine*, QJSEngine*)>)'
```
It's quite strange, because I could not identify any issue in this code.

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+}
+
+QGrpcChannelSubscription::~QGrpcChannelSubscription() {
+	cancel();
```
Tab instead of spaces

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+    return call.status;
+}
+
+void QGrpcChannelPrivate::subscribe(QGrpcSubscription *subscription, const QString &service, QAbstractGrpcClient *client)  {
```
New line before '{' is preferred according project coding style

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+    call->start();
+}
+
+QGrpcStatus QGrpcChannelPrivate::call(const QString &method, const QString &service, const QByteArray &args, QByteArray &ret) {
```
New line before '{' is preferred according project coding style

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+QGrpcChannelPrivate::~QGrpcChannelPrivate() {
+}
+
+void QGrpcChannelPrivate::call(const QString &method, const QString &service, const QByteArray &args, QGrpcAsyncReply *reply) {
```
New line before '{' is preferred according project coding style

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+    return ::grpc::Status::OK;
+}
+
+QGrpcChannelSubscription::QGrpcChannelSubscription(grpc::Channel *channel, const QString &method, const QByteArray &data, QObject *parent) : QObject(parent) {
```
New line before '{' is preferred according project coding style

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+        QObject::disconnect(*abortConnection);
+        QObject::disconnect(*clientConnection);
+
+		sub->cancel();
```
Tabs instead of spaces

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+        QObject::disconnect(*abortConnection);
+        QObject::disconnect(*clientConnection);
+
+		sub->cancel();
```
Tabs instead of spaces

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+            case QGrpcStatus::Ok:
+                break;
+            default:
+                qDebug()<<sub->status.code();
```
Please use qProtoDebug instead. Add spaces arround <<

rapgenic replied:
That was actually supposed not to be there...

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+        QObject::disconnect(*abortConnection);
+        QObject::disconnect(*clientConnection);
+
+        switch (sub->status.code()) {
```
It seems 'if else' might be better here

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+
+namespace QtProtobuf {
+
+static inline grpc::Status parseByteBuffer(const grpc::ByteBuffer &buffer, QByteArray *data) {
```
Probably non-constant reference to QByteArray &data is better in this case?

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+    return status;
+}
+
+static inline grpc::Status parseQByteArray(const QByteArray &bytearray, grpc::ByteBuffer *buffer) {
```
Probably non-constant reference to grpc:ByteBuffer &buffer is better in this case?

semlanik replied:
Overall function might be void as far as I could see. It always returns ::grpc::Status::OK and result is never checked when function called.

semlanik commented on src/grpc/qgrpcchannel.cpp:
```diff
@@ -0,0 +1,343 @@
+    std::vector<grpc::Slice> slices;
+    auto status = buffer.Dump(&slices);
+
+    for (auto slice : slices) {
```
Does it really make sense to continue parsing in case if status is not OK?

semlanik commented:
@rapgenic , sorry for my pedantic review style, it's easier to see issues in someone else's code than in own :) Anyway commit looks good for me, looking forward for compilation issue fix, probably it's related to ci's compiler version. I checked locally and don't observe this compilation issue.

rapgenic commented:
Please tell me if there is something else you need me to do! 

When you think that everything is ok I can squash all the commits into a single one for a cleaner history

semlanik commented:
@rapgenic , you may squash your commits and I will merge this pull request. Thank you for this great contribution!

rapgenic commented:
You're welcome, and thank you for your support!
